src/node_v8.cc

```diff
@@ -20,25 +20,34 @@ using v8::Uint32;
 using v8::V8;
 using v8::Value;
 
+#define HEAP_STATISTICS_PROPERTIES(V)                                         \
+  V(0, total_heap_size, kTotalHeapSizeIndex);                                 \
+  V(1, total_heap_size_executable, kTotalHeapSizeExecutableIndex);            \
+  V(2, total_physical_size, kTotalPhysicalSizeIndex);                         \
+  V(3, used_heap_size, kUsedHeapSizeIndex);                                   \
+  V(4, heap_size_limit, kHeapSizeLimitIndex);
+
+#define kHeapStatisticsBufferLength 5
+#define kHeapStatisticsBufferType v8::kExternalUint32Array
 
 void GetHeapStatistics(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
+  CHECK(args.Length() == 1 && args[0]->IsObject());
+
   Isolate* isolate = args.GetIsolate();
   HeapStatistics s;
   isolate->GetHeapStatistics(&s);
-  Local<Object> info = Object::New(isolate);
-  // TODO(trevnorris): Setting many object properties in C++ is a significant
-  // performance hit. Redo this to pass the results to JS and create/set the
-  // properties there.
-#define V(name)                                                               \
-  info->Set(env->name ## _string(), Uint32::NewFromUnsigned(isolate, s.name()))
-  V(total_heap_size);
-  V(total_heap_size_executable);
-  V(total_physical_size);
-  V(used_heap_size);
-  V(heap_size_limit);
+  Local<Object> obj = args[0].As<Object>();
+  uint32_t* data = static_cast<uint32_t*>(obj->GetIndexedPropertiesExternalArrayData());
+
+  CHECK_NE(data, nullptr);
+  ASSERT(obj->GetIndexedPropertiesExternalArrayDataType() == kHeapStatisticsBufferType);
+  ASSERT(obj->GetIndexedPropertiesExternalArrayDataLength() == kHeapStatisticsBufferLength);
```

Comment:
Style: long lines, please wrap at 80 columns.

I realize we don't have ASSERT_EQ yet.  Can you update this when #529 lands?